### PR TITLE
Changed caching method

### PR DESCRIPTION
**Pull request by trapacska**

### CODE REVIEW

File: .gows.user.yml [hunk 1]
```diff
@@ -0,0 +1 @@
+sync_mode: symlink
```
Comment (godrei) [inline .gows.user.yml]: gitignore this file

File: bitrise.yml [hunk 2]
```diff
@@ -10,24 +10,17 @@ app:
 
 workflows:
   test:
+    before_run:
+    - go-tests
     steps:
     - script:
-        title: Go Test
+        title: Download assets
```
Comment (godrei) [inline bitrise.yml]: do this in the _tmp dir

File: main.go [hunk 3]
```diff
@@ -1,732 +1,717 @@
...
 	}
-	defer func() {
-		if err := f.Close(); err != nil {
-			log.Printf(" (!) Failed to close file (%s) after creating its checksum, error: %s", pth, err)
-		}
-	}()
 
-	if _, err := io.Copy(fileHasher, f); err != nil {
-		return nil, fmt.Errorf("Failed create hash of file, error: %s", err)
+	if _, err := io.CopyN(cacheModel.TarWriter, bytes.NewReader(filePathMapBytes), filePathMapSize); err != nil && err != io.EOF {
```
Comment (godrei) [inline main.go]: should not we use: `io.Copy(cacheModel.TarWriter, bytes.NewReader(filePathMapBytes))`, if we copy the whole filePathMapBytes

File: main.go [hunk 4]
```diff
@@ -1,732 +1,717 @@
...
-			}
-		} else {
-			// prefix
-			if strings.HasPrefix(aPth, anIgnorePth) {
-				// log.Printf(" [IGNORE:prefix] %s (%s)", aPth, anIgnorePth)
-				return true
-			}
-		}
+// GenerateCacheInfoMAP ...
+func (cacheModel *CacheModel) GenerateCacheInfoMAP() (map[string]string, error) {
```
Comment (godrei) [inline main.go]: `GenerateCacheInfoMAP` -> `GenerateCacheInfoMap`

File: main.go [hunk 5]
```diff
@@ -1,732 +1,717 @@
...
+					file, err := os.Open(path)
+					if err != nil {
+						return err
+					}
 
-				if gIsDebugMode {
-					log.Printf(" * fileFingerprintSource (%s): %#v", aPath, fileFingerprintSource)
-				}
-				fingerprintMeta[aPath] = FingerprintMetaModel{FingerprintSource: fileFingerprintSource}
+					_, err = io.CopyN(cacheModel.TarWriter, file, info.Size())
```
Comment (godrei) [inline main.go]: if io.CopyN fails the file will not be closed, use defer, after `file, err := os.Open(path)`

File: main.go [hunk 6]
```diff
@@ -1,732 +1,717 @@
...
-			fileInfo, isExist, err := pathutil.PathCheckAndInfos(aPath)
+			indicatorFileInfo, indicatorFilePathExists, err := pathutil.PathCheckAndInfos(indicatorFilePath)
 			if err != nil {
-				log.Printf(" "+colorstring.Yellow("(!) Skipping")+" (%s): Failed to check the specified path: %s", aOrigPthItm.Path, err)
-				continue
+				return err
+			}
+			if !indicatorFilePathExists {
+				log.Errorf("Indicator file doesn't exists: %s", cleanPath)
+				os.Exit(1)
```
Comment (godrei) [inline main.go]: os.Exit, not retrun?

File: main.go [hunk 7]
```diff
@@ -1,732 +1,717 @@
...
-			archiveCopyRsyncParams = []string{"-avhP", absItemPath, filepath.Join(cacheContentDirPath, itemRelPathInArchive)}
-		}
+	return nil
+}
 
-		cacheInfo.Contents = append(cacheInfo.Contents, CacheContentModel{
-			DestinationPath:       aPath,
-			RelativePathInArchive: itemRelPathInArchive,
-		})
+func main() {
```
Comment (godrei) [inline main.go]: move this to the end of the file

File: main.go [hunk 8]
```diff
@@ -781,159 +763,56 @@ func getCacheUploadURL(cacheAPIURL string, fileSizeInBytes int64) (string, error
...
-		fmt.Println()
-		time.Sleep(3000 * time.Millisecond)
-		return _tryToUploadArchive(uploadURL, archiveFilePath)
-	}
-	return nil
-}
-
-func main() {
-	stepParams, err := CreateStepParamsFromEnvs()
+func _tryToUploadArchive(uploadURL string, archiveFilePath string) error {
```
Comment (godrei) [inline main.go]: _tryToUploadArchive -> tryToUploadArchive

File: main.go [hunk 9]
```diff
@@ -1,732 +1,717 @@
...
 	}
 
-	if resp.StatusCode != 200 {
-		log.Printf("=> Upload response: %s", responseBytes)
-		return fmt.Errorf("Failed to upload file, response code was: %d", resp.StatusCode)
-	}
-	if gIsDebugMode {
-		log.Printf("=> Upload response code: %d", resp.StatusCode)
+	err = f.Close()
+	if err != nil {
```
Comment (godrei) [inline main.go]: if err != nil { -> `if err := f.Close(); err != nil {`

File: main.go (hunk 8 above)
Comment (godrei) [reply main.go]: please review the whole function, why is `isFileCloseRequired` required?

**Comment by Dschee** [comment]
Version 2.0.0 breaks our Carthage cache. I can see that you "changed your caching method" but there are no release notes on "what" you changed. Is there any document where we can see how we need to migrate to this new version?

**Comment by viktorbenei** [comment]
@Dschee please create a separate issue, for tracking & notification purposes

**Comment by Dschee** [comment]
I just meant to ask a question. But I think I know the answer niw. ^^

**Comment by viktorbenei** [comment]
Glad to hear @Dschee :)

**Comment by Dschee** [comment]
I think the answer is "No, we don't have such a document". Not sure if you understood that. ^^

**Comment by viktorbenei** [comment]
@Dschee I might have missed something, but what changed in the PR: https://github.com/bitrise-steplib/steps-cache-push/pull/11/files

What changed in 2.0.0 (between 1.1.4 and 2.0.0): https://github.com/bitrise-steplib/steps-cache-push/compare/1.1.4...2.0.0

**Pull request merged**